File: algorithms/ekm.m
function [idx, C, numit, W, U, sumd, D, J] = ekm(X, K, options)
%
% Version 1
% Created at June, 2024
% Last modified at June, 2025
% Author: Yudong He
% Email: yhebh@connect.ust.hk
% Tested Matlab version: R2022a
%% Description:
%
% Inputs:
% X : N-by-P numeric matrix. Data matrix containing N samples with P features.
% K : positive integer. Number of clusters.
% options : Struct containing optional parameters that govern the algorithm's behavior.
% - Distance: 'euclidean' (default) | 'manhattan'
%   'euclidean': euclidean distance or l2 norm.
%   'manhattan': manhattan distance or l1 norm.
%
% - alpha: double (default 0.5) | "dvariance". Smoothing parameter or method to calculate the smoothing parameter.
%   'dvariance': data variance.
%
% - MaxIter: 500 (default) | positive integer. Maximum number of iterations.
%
% - Eta: 1e-3 | positive double. Tolerance for convergence
% 
% - Replicates: 1 (default) | positive integer. Replicate number. The replication with the lowest objetive value will be chosen as the final outcome.
%
% - Start: 'plus' (default) | numeric matrix. Method for initializing centroid or K-by-P matrix containing initial centroids
%   'plus': the k-means++ algorithm
%
% Outputs:
% idx : N-by-1 numeric column vector. The cluster indices for each sample.
% C : K-by-P numeric matrix. Estimated centroids of the clusters.
% W : N-by-K numeric matrix. Weights for each sample and cluster.
% sumd : K-by-1 numeric column vector. Within-cluster sums of point-to-centroid distances.
% D : N-by-K numeric matrix. Distances from each data point to each centroid.
% J : scalar number. The lowest loss value over replications.
% 
% Note: This function implements an enhanced K-means algorithm, named equilibrium K-means (EKM). EKM is robust to imbalanced data clustering.
%
%% Example:
% [idx,C] = ekm(X,K) returns estimated indices and centroids given instances X and the number of clusters K.
% [idx,C] = ekm(X,K,"alpha",1) specifies the only one hyperparameter, alpha, in EKM.
% [idx,C] = ekm(X,K,"Replicate",100) specifies the number
% of replications. The K-means++ is used to intialize centroid for each
% replication

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Reference
% He Yudong, An Equilibrium Approach to Clustering: Surpassing Fuzzy C-Means
% on Imbalanced Data, IEEE Transactions on Fuzzy Systems, 2025.

% He Yudong, Imbalanced Data Clustering Using Equilibrium K-Means, arXiv, 2024.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Copyright 2024 Yudong He
% This software is distributed under the terms of the GNU Public License
% version 3 (http://www.gnu.org/licenses/gpl.txt)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    
    %% Arguments validation
    arguments
        X (:,:) double {mustBeNonempty, mustBeFinite, mustBeReal}
        K double {mustBeInteger, mustBePositive}
        options.metric char {mustBeText} = 'euclidean'
        options.alpha = 0.5
        options.MaxIter double {mustBeInteger, mustBePositive} = 500
        options.Eta double {mustBePositive} = 1e-3 % for convergence
        options.Replicates double {mustBeInteger, mustBePositive} = 1
        options.Start = 'plus'
    end
    
    %% Initialize necessary variables
    if isnumeric(options.Start)
        R = size(options.Start, 3);
    else
        R = options.Replicates;
    end
   
    % Define distance function based on user selection
    switch options.metric
        case 'euclidean'
            dist = @euclidean;
        case 'manhattan'
            dist = @manhattan;
        otherwise
            error('Unsupported distance metric specified.');
    end
    
    % Handle the alpha parameter based on its type
    if ischar(options.alpha)
        switch options.alpha
            case 'dvariance'
                options.alpha = 2/(mean(dist(X,mean(X)).^2)); % determine alpha based on data variance
            otherwise
                error('Unsupported alpha type specified.');
        end
    end
    
    %% Prepare for multiple replicates
    J_set = zeros(R, 1);
    C_set = zeros(K, size(X, 2), R); % Preallocate for centroids
    numit_set = zeros(R, 1);
    
    %% Start replicates
    for r = 1:R
        % Initialize centroid based on specified method
        if isnumeric(options.Start)
            C = squeeze(options.Start(:, :, r)); % K-by-P matrix
        else
            switch options.Start
                case 'plus' % K-means++ initialization
                    C = kmeans_plus_init(X, K);
                otherwise
                    error('Unsupported initialization method specified.');
            end
        end
        
        %% Find Centroids and Assign Weights
        C_old = C; % Initialize old centroids
        it = 1;    % Iteration counter
        while true
            D = dist(X, C); % Calculate distances
            W = calc_weight(D, options.alpha); % Calculate weights
            
            % Update centroids
            for k = 1:K
                sum_Wk = sum(W(:, k));
                if sum_Wk == 0
                    sum_Wk = eps; % Prevent NaN for empty clusters
                end
                C(k, :) = sum(W(:, k) .* X, 1) / sum_Wk; % Update centroid
            end
            
            % Check for convergence
            if norm(C - C_old, 'fro') / norm(C, 'fro') < options.Eta
                break; % Convergence achieved
            end
            
            % Check if maximum iterations reached
            if it >= options.MaxIter
                fprintf('Failed to converge in %d iterations during replicate %d for EKM with %d clusters\n', ...
                        options.MaxIter, r, K);
                break;
            else
                C_old = C; % Update old centroid for next iteration
                it = it + 1; % Increment iteration count
            end
        end
        
        %% Calculate objectives
        D = dist(X, C);
        J = sum(sum(D.^2 .* exp(-options.alpha * D.^2), 2) ./ (sum(exp(-options.alpha * D.^2), 2) + eps));
        J_set(r) = J;
        C_set(:, :, r) = C; % Store centroids
        numit_set(r) = it; % Store number of iterations
    end
    
    %% Final Results Selection
    [~, best_id] = min(J_set);
    J = J_set(best_id);
    C = C_set(:, :, best_id);
    numit = numit_set(best_id);
    D = dist(X, C); % Final distance calculation
    
    % Assign cluster indices
    [~, idx] = min(D, [], 2);
    
    % Calculate the within-cluster sums of point-to-centroid distances
    sumd = arrayfun(@(k) sum(D(idx == k, k)), 1:K)'; % Vectorized for each cluster
    
    % Compute weights
    W = calc_weight(D, options.alpha);
    % Compute membership
    U = exp(-options.alpha .* D.^2) ./ (sum(exp(-options.alpha .* D.^2), 2) + eps); % Normalized memberships
end

function C = kmeans_plus_init(X, K)
    % K-means++ initialization for centroid selection
    tmp=X'; % P-by-N matrix
    C = tmp(:,1+round(rand*(size(tmp,2)-1))); % randomly select the first centroid
    L = ones(1,size(tmp,2));
    for i = 2:K
        D = tmp-C(:,L);
        D = cumsum(sqrt(dot(D,D,1)));
        if D(end) == 0, C(:,i:K) = tmp(:,ones(1,K-i+1)); return; end
        C(:,i) = tmp(:,find(rand < D/D(end),1));
        [~,L] = max(bsxfun(@minus,2*real(C'*tmp),dot(C,C,1).'));
    end
    C=C'; % K-by-P matrix
end

function D = euclidean(X, C)
    % Euclidean distance computation
    % X: N-by-P matrix; C: K-by-P matrix
    N=size(X,1);
    K=size(C,1);
    D = zeros(N, K);
    for k = 1:K
        D(:, k) = vecnorm(X - C(k, :), 2, 2); % Calculate euclideaan distance vector
    end
end

function D = manhattan(X, C)
    % Manhattan distance computation
    % X: N-by-P matrix; C: K-by-P matrix
    N=size(X,1);
    K=size(C,1);
    D = zeros(N, K);
    for k = 1:K
        D(:, k) = sum(abs(X - C(k, :)), 2); % Calculate euclideaan distance vector
    end
end

function W = calc_weight(D, alpha)
    % Calculate weights based on distances D and smoothing parameter alpha
    K=size(D,2);
    J=sum(D.^2.*exp(-alpha*D.^2),2)./(sum(exp(-alpha*D.^2),2)+eps); % objectives contributed by N points individually
    W=exp(-alpha*D.^2)./repmat(sum(exp(-alpha*D.^2),2)+eps,1,K).*(1-alpha*(D.^2-repmat(J,1,K)));
    % prevent all 0 membership because of numerical precision
    zero_idx=find(sum(W,2)==0);
    [~,pos]=min(D(zero_idx,:),[],2);
    W(zero_idx,:) = (pos==1:K);
end

File: python/ekm.py
import numpy as np
def ekm(X, K, 
        Distance='sqeuclidean', 
        alpha=0.5, 
        MaxIter=500, 
        Eta=1e-3, 
        Replicates=1, 
        Start='plus'):
    """
    Enhanced K-means (Equilibrium K-means, EKM).
    Parameters
    ----------
    X : ndarray of shape (N, P)
        Data matrix containing N samples with P features.
    K : int
        Number of clusters.
    Distance : str, default 'sqeuclidean'
        'sqeuclidean' or 'cosine'
    alpha : float or str, default 0.5
        Smoothing parameter or 'dvariance'
    MaxIter : int, default 500
        Maximum number of iterations.
    Eta : float, default 1e-3
        Tolerance for convergence.
    Replicates : int, default 1
        Number of replications. The replication with lowest objective is retained.
    Start : str or ndarray, default 'plus'
        Initialization method ('plus') or K x P ndarray for user-defined centroids.
    Returns
    -------
    idx : ndarray of shape (N,)
        Cluster indices per sample.
    C : ndarray of shape (K, P)
        Centroids of clusters.
    numit : int
        Number of iterations performed.
    W : ndarray of shape (N, K)
        Weights per sample per cluster.
    U : ndarray of shape (N, K)
        Membership values.
    sumd : ndarray of shape (K,)
        Sum of distances within each cluster.
    D : ndarray of shape (N, K)
        Distance matrix.
    J : float
        Objective function value.
    """
    N, P = X.shape
    # Choose distance function
    if Distance == 'sqeuclidean':
        dist = sqeuclidean
    elif Distance == 'cosine':
        dist = cosine
    else:
        raise ValueError("Unsupported distance metric.")
    # Handle alpha
    if isinstance(alpha, str):
        if alpha == 'dvariance':
            alpha = 2.0 / np.mean(dist(X, np.mean(X, axis=0).reshape(1, -1)))
        else:
            raise ValueError("Unsupported alpha option.")
    # Replicates handling
    if isinstance(Start, np.ndarray):
        R = Start.shape[0] // K
    else:
        R = Replicates
    J_set = np.zeros(R)
    C_set = np.zeros((R, K, P))
    numit_set = np.zeros(R, dtype=int)
    for r in range(R):
        # initialization
        if isinstance(Start, np.ndarray):
            C = Start[r*K:(r+1)*K, :]
        else:
            if Start == 'plus':
                C = kmeans_plus_init(X, K)
            else:
                raise ValueError("Unsupported initialization method.")
        C_old = C.copy()
        it = 1
        while True:
            D = dist(X, C)
            W = calc_weight(D, alpha)
            # update centroids
            for k in range(K):
                sum_wk = np.sum(W[:, k])
                if sum_wk == 0:
                    sum_wk = np.finfo(float).eps
                C[k, :] = np.dot(W[:, k], X) / sum_wk
            # convergence check
            if np.linalg.norm(C - C_old, 'fro') / np.linalg.norm(C, 'fro') < Eta:
                break
            if it >= MaxIter:
                print(f"Failed to converge in {MaxIter} iterations during replicate {r+1} for EKM with {K} clusters")
                break
            else:
                C_old = C.copy()
                it += 1
        D = dist(X, C)
        J = np.sum(np.sum(D * np.exp(-alpha*D), axis=1) / (np.sum(np.exp(-alpha*D), axis=1) + np.finfo(float).eps))
        J_set[r] = J
        C_set[r] = C
        numit_set[r] = it
    best_id = np.argmin(J_set)
    J = J_set[best_id]
    C = C_set[best_id]
    numit = numit_set[best_id]
    D = dist(X, C)
    idx = np.argmin(D, axis=1)
    sumd = np.array([np.sum(D[idx == k, k]) for k in range(K)])
    W = calc_weight(D, alpha)
    U = np.exp(-alpha * D) / (np.sum(np.exp(-alpha * D), axis=1, keepdims=True) + np.finfo(float).eps)
    return idx, C, numit, W, U, sumd, D, J
def kmeans_plus_init(X, K):
    """K-means++ initialization"""
    N, P = X.shape
    C = np.empty((K, P))
    idx = np.random.randint(N)
    C[0] = X[idx]
    for k in range(1, K):
        D = np.min(sqeuclidean(X, C[:k]), axis=1)
        probs = D / np.sum(D)
        idx = np.random.choice(N, p=probs)
        C[k] = X[idx]
    return C
def sqeuclidean(X, C):
    """Squared Euclidean distance"""
    return np.sum((X[:, None, :] - C[None, :, :])**2, axis=2)
def cosine(X, C):
    """Cosine distance"""
    X_norm = X / (np.linalg.norm(X, axis=1, keepdims=True) + np.finfo(float).eps)
    C_norm = C / (np.linalg.norm(C, axis=1, keepdims=True)[:, None] + np.finfo(float).eps)
    return 1 - np.dot(X_norm, C_norm.T)
def calc_weight(D, alpha):
    """Weight calculation"""
    N, K = D.shape
    J = np.sum(D * np.exp(-alpha * D), axis=1) / (np.sum(np.exp(-alpha * D), axis=1) + np.finfo(float).eps)
    W = np.exp(-alpha * D) / (np.sum(np.exp(-alpha * D), axis=1, keepdims=True) + np.finfo(float).eps) \
        * (1 - alpha * (D - J[:, None]))
    zero_idx = np.where(np.sum(W, axis=1) == 0)[0]
    for i in zero_idx:
        pos = np.argmin(D[i])
        W[i] = np.eye(1, K, pos)
    return W